src/data_structure/circular_queue.rs

```diff
@@ -0,0 +1,105 @@
+pub struct Queue<T> {
+    data: Vec<T>,
+    index: usize,
+    front: usize,
+    capacity: usize,
+}
+
+impl<T> Queue<T> {
+    // Returns a new Queue object with given capacity
+    pub fn with_capacity(capacity: usize) -> Self {
+        Self {
+            data: Vec::with_capacity(capacity),
+            index: 0,
+            front: 0,
+            capacity,
+        }
+    }
+
+    // Adds the element to the circular queue
+    // Once capacity is reached, new items will overwrite old ones.
+    pub fn enqueue(&mut self, e: T) {
+        if self.len() < self.capacity() {
+            self.data.push(e);
+        } else {
+            self.data[self.index] = e;
+        }
+        self.index = (self.index + 1) % self.capacity();
+    }
+
+    // Returns an Option with first element of the list,
+    // if it is empty returns None
+    pub fn dequeue(&mut self) -> Option<T> {
+        if self.is_empty() {
+            return None;
+        }
+        let ele = self.data.swap_remove(self.front as usize);
```

Comment:
I am not sure if we want to do this, since `swap_remove` removes the element at `self.front` and replaces it with the last element of the array, so I think it places the element at the end of queue to the front, and If after this, we `enqueue` other elements, they will probably be out-of-order.
We should also add more tests to cover various scenarios possible :)

Reply:
`enqueue` always places the element, so it's never out of bounds. It will overwrite the element at the front **always**. To have O(1) operatoins in a vec, this is what I used. If you still want the change, I'll have to rewrite it in VecDeque.

Reply:
Okay, can you check for this test is it passes?

```
let mut queue: Queue<i32> = Queue::with_capacity(5);
queue.enqueue(1);
queue.enqueue(2);
queue.enqueue(3);
queue.dequeue(); // this should remove 1
assert_eq!(queue.dequque().unwrap(), 2)

```

Reply:
Yes, I'm working on adding test for the dequeue elements.

Reply:
I am still thinking how is it working, can you try this?
```
        let mut queue: Queue<i32> = Queue::with_capacity(5);
        assert!(queue.dequeue().is_none());
        queue.enqueue(1);
        queue.enqueue(2);
        queue.enqueue(3);
        queue.enqueue(4);
        assert_eq!(queue.dequeue(), Some(1));
        assert_eq!(queue.dequeue(), Some(2));
        assert_eq!(queue.dequeue(), Some(3));
        assert_eq!(queue.dequeue(), Some(4));
```

Reply:
This fails, it fails at `Some(3)`, due to the new vec having a different length the front is miscalculated.
What do you suggest I do ?

Reply:
We can chat about it on Gitter :)